### ISSUE TITLE
constexpr can't handle vector types

### ISSUE THREAD

**Issue by RKSimon** [llvm/llvm-project#41806]
|  |  |
| --- | --- |
| Bugzilla Link | [42461](https://llvm.org/bz42461) |
| Version | unspecified |
| OS | Windows NT |
| Blocks | llvm/llvm-project#30794  |
| CC | @erichkeane,@mshockwave,@zygoloid |
| Fixed by commit(s) | b30c16670e42 |

## Extended Description 
Godbolt: https://godbolt.org/z/l-RaB7

We are unable to use __vector_size or ext_vector_type defined types in constexpr functions.
```
typedef int __v4si __attribute__((__vector_size__(16)));

constexpr int sum_i32(__v4si x) {
  return x[0] + x[1] + x[2] + x[3];
}

constexpr __v4si add_i32(__v4si x, __v4si y) {
  return x + y;
}
```
clang -g0 -O3 
```
<source>:3:18: error: constexpr function never produces a constant expression [-Winvalid-constexpr]

constexpr __v4si add_i32(__v4si x, __v4si y) {
                 ^

<source>:4:12: note: subexpression not valid in a constant expression
  return x + y;

           ^

<source>:7:15: error: constexpr function never produces a constant expression [-Winvalid-constexpr]

constexpr int sum_i32(__v4si x) {

              ^

<source>:8:10: note: subexpression not valid in a constant expression
  return x[0] + x[1] + x[2] + x[3];
```

**Comment by yronglin** [comment]
IIUC, this issue has already fixed. https://godbolt.org/z/rccbYzYrc

**Comment by RKSimon** [comment]
> IIUC, this issue has already fixed. https://godbolt.org/z/rccbYzYrc

I think this was addressed by #101126 / 77534291fcbd2c784c54e39a60895e4f60f19742
